src/support/nonce/manager.js

```diff
@@ -68,48 +68,65 @@ class NonceManager {
   async getNonceForAccount(address) {
     logger.debug(`Requesting nonce for address ${address}`);
 
-    // Retrieve stored nonces for the provided account.
-    let nonces = await this.store.get(address);
-    // Retrieve current transaction count and transaction pool state for the provided account.
-    const [txCount, { pending, queued }] = await Promise.all([
-      this.getTransactionCount(address),
-      this.inspectTxPool(address)
-    ]);
-
-    // Keep nonces which are not mined yet
-    // and release nonces which values are below the account tx count (i.e. lowest possible value).
-    nonces = _.pickBy(nonces, (value, nonce) => {
-      if (nonce >= txCount) return true;
-      logger.debug(`Account '${address}', nonce released: ${Number(nonce)}`);
-      return false;
-    });
-
-    // Get all known transactions by combining local cache with data from tx pool.
-    const knownTransactions = _.assign({}, nonces, pending, queued);
-
-    // Get all used nonces.
-    const usedNonces = _.keys(knownTransactions);
-    if (usedNonces.length) {
-      logger.debug(`Account '${address}', used nonces: ${usedNonces.join(', ')}`);
+    // First acquire lock to avoid race conditions when modifying the list of used nonces
+    try {
+      await this.store.lock(address);
+    } catch (error) {
+      logger.error(`Error during acquiring lock for ${address}: ${error.message}`, mapError(error));
+      throw error;
     }
 
-    // Calculate max used nonce.
-    const maxUsedNonce = usedNonces.reduce((a, b) => Math.max(a, b), txCount);
+    try {
+      // Retrieve current transaction count and transaction pool state for the provided account.
+      const [txCount, { pending, queued }] = await Promise.all([
+        this.getTransactionCount(address),
+        this.inspectTxPool(address)
+      ]);
+
+      // Retrieve stored nonces for the provided account.
+      let nonces = await this.store.get(address);
+
+      // Keep nonces which are not mined yet
+      // and release nonces which values are below the account tx count (i.e. lowest possible value).
+      nonces = _.pickBy(nonces, (value, nonce) => {
```

Comment:
I think this is nitpicking but I don't like so much to reassign values like that. Maybe we could call it notMinedNonces?

Reply:
This is a code piece that I haven't touched in this PR. I just wrapped it with try{}catch, that's why it's shown as modified.
I think this code is rather hard to read and understand, I will try to refactor it a bit right after I will cove nonce manager with proper unit tests. Definitely in some other PR, so let's close eyes on this one for now.

Reply:
Ok, I'm fine with that.